src/lib/support/ASN1.h

```diff
@@ -0,0 +1,217 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2013-2017 Nest Labs, Inc.
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines types and objects for reading and writing
+ *      Abstract Syntax Notation One (ASN.1) encoded data.
+ *
+ */
+
+#ifndef ASN1_H_
+#define ASN1_H_
+
+#include <support/DLLUtil.h>
+
+#include "ASN1Error.h"
+
+namespace chip {
+namespace TLV {
+class TLVReader;
+}
+} // namespace chip
+
+/**
+ *   @namespace chip::ASN1
+ *
+ *   @brief
+ *     This namespace includes all interfaces within CHIP for
+ *     working with Abstract Syntax Notation One (ASN.1).
+ */
+
+namespace chip {
+namespace ASN1 {
+
+#include <support/ASN1OID.h>
+
+enum ASN1TagClasses
+{
+    kASN1TagClass_Universal       = 0x00,
+    kASN1TagClass_Application     = 0x40,
+    kASN1TagClass_ContextSpecific = 0x80,
+    kASN1TagClass_Private         = 0xC0
+};
+
+enum ASN1UniversalTags
+{
+    kASN1UniversalTag_Boolean         = 1,
+    kASN1UniversalTag_Integer         = 2,
+    kASN1UniversalTag_BitString       = 3,
+    kASN1UniversalTag_OctetString     = 4,
+    kASN1UniversalTag_Null            = 5,
+    kASN1UniversalTag_ObjectId        = 6,
+    kASN1UniversalTag_ObjectDesc      = 7,
+    kASN1UniversalTag_External        = 8,
+    kASN1UniversalTag_Real            = 9,
+    kASN1UniversalTag_Enumerated      = 10,
+    kASN1UniversalTag_UTF8String      = 12,
+    kASN1UniversalTag_Sequence        = 16,
+    kASN1UniversalTag_Set             = 17,
+    kASN1UniversalTag_NumericString   = 18,
+    kASN1UniversalTag_PrintableString = 19,
+    kASN1UniversalTag_T61String       = 20,
+    kASN1UniversalTag_VideotexString  = 21,
+    kASN1UniversalTag_IA5String       = 22,
+    kASN1UniversalTag_UTCTime         = 23,
+    kASN1UniversalTag_GeneralizedTime = 24,
+    kASN1UniversalTag_GraphicString   = 25,
+    kASN1UniversalTag_VisibleString   = 26,
+    kASN1UniversalTag_GeneralString   = 27,
+    kASN1UniversalTag_UniversalString = 28
+};
+
+class ASN1UniversalTime
+{
+public:
+    uint16_t Year;
+    uint8_t Month;
+    uint8_t Day;
+    uint8_t Hour;
+    uint8_t Minute;
+    uint8_t Second;
+};
+
+class DLL_EXPORT ASN1Reader
+{
+public:
+    void Init(const uint8_t * buf, uint32_t len);
+
+    uint8_t Class;
```

Comment:
Could we make the data members non-public?

Reply:
done